In [1]:
import pandas as pd
import numpy as np
import urllib

ASSIGNING WIKIPEDIA URL IN A VARIABLE url

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

UTILIZING BEAUTIGULSOUP LIBRARY FOR WEB SCRAPING

In [3]:
page = urllib.request.urlopen(url)

In [4]:
from bs4 import BeautifulSoup

In [5]:
soup = BeautifulSoup(page, "lxml")

In [6]:
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"62260b64-7a66-48f9-a619-e2fff4e46fbe","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":967921175,"wgRevisionId":967921175,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

ASSINGING HTML TABLE TO VARABLE table

In [8]:
table = soup.find('table', class_= 'wikitable sortable')
table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td>

CREATING 3 EMPTY LISTS AND APPENDING THE SCRAPED DATA TO THE LISTS, TR AND TD ARE THE TAGS BETWEEN ROWS AND AND INDIVIDUAL VALUES IN THE SCRAPED DATA

In [9]:
a= []
b=[]
c=[]
for row in table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        a.append(cells[0].find(text=True))
        b.append(cells[1].find(text=True))
        c.append(cells[2].find(text=True))

REMOVING NEW LINE CHARACTER FROM ALL THE VALUES

In [10]:
a1 =[]
b1 =[]
c1 =[]
import re
for x in a:
    a1.append(re.sub('\n', '', x))
for x in b:
    b1.append(re.sub('\n', '', x))
for x in c:    
    c1.append(re.sub('\n', '', x))    

df=pd.DataFrame(a1,columns=['postalcode'])
df['borough']=b1
df['neighborhood']=c1
df.head(2)

postalcode       borough  neighborhood
0        M1A  Not assigned  Not assigned
1        M2A  Not assigned  Not assigned

DROPPING ROWS WHERE BOROUGH IS 'NOT ASSIGNED'

In [11]:
df.drop(df.loc[df['borough']=='Not assigned'].index, inplace=True)

LOOK AT THE DATAFRAME HEAD

In [12]:
df.head()

postalcode           borough                                 neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [13]:
df = df.reset_index()

In [14]:
df.drop(['index'], axis = 1, inplace = True)

RESETTING INDEX OF DATAFRAME

In [15]:
df.head()

postalcode           borough                                 neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

CHECKING IF ALL NEIGHBORHOOD HAVE SOME VALUE

In [16]:
(df['neighborhood']=='Not assigned').value_counts()

False    103
Name: neighborhood, dtype: int64

In [17]:
df.shape

(103, 3)

ALL NEIHGBORHOOD HAVE SOME VALUE, DATAFRMAE HAS 103 ROWS AND 3 COLUMNS

In [18]:
# The code was removed by Watson Studio for sharing.

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [19]:
df_data_1.shape

(103, 3)

In [20]:
df.shape

(103, 3)

In [21]:
df_data_1.rename(columns = {'Postal Code' : 'postalcode'}, inplace = True)

In [22]:
df.head()

postalcode           borough                                 neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [23]:
df_data_1.head()

postalcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [24]:
df1 = pd.merge(left=df, right=df_data_1, left_on='postalcode', right_on='postalcode')

In [25]:
df1.head()

postalcode           borough                                 neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [26]:
df1.shape

(103, 5)

In [27]:
for x in df1['borough']:
    if df1['borough'].str.contains('Toronto').any():
        print (df1['borough'])
    else:
        print('no')

0            North York
1            North York
2      Downtown Toronto
3            North York
4      Downtown Toronto
5             Etobicoke
6           Scarborough
7            North York
8             East York
9      Downtown Toronto
10           North York
11            Etobicoke
12          Scarborough
13           North York
14            East York
15     Downtown Toronto
16                 York
17            Etobicoke
18          Scarborough
19         East Toronto
20     Downtown Toronto
21                 York
22          Scarborough
23            East York
24     Downtown Toronto
25     Downtown Toronto
26          Scarborough
27           North York
28           North York
29            East York
             ...       
73      Central Toronto
74      Central Toronto
75         West Toronto
76          Mississauga
77            Etobicoke
78          Scarborough
79      Central Toronto
80     Downtown Toronto
81         West Toronto
82          Scarborough
83      Central 

In [34]:
df1['borough'].str.contains('Toronto').to_frame()

borough
0      False
1      False
2       True
3      False
4       True
5      False
6      False
7      False
8      False
9       True
10     False
11     False
12     False
13     False
14     False
15      True
16     False
17     False
18     False
19      True
20      True
21     False
22     False
23     False
24      True
25      True
26     False
27     False
28     False
29     False
..       ...
73      True
74      True
75      True
76     False
77     False
78     False
79      True
80      True
81      True
82     False
83      True
84      True
85     False
86      True
87      True
88     False
89     False
90     False
91      True
92      True
93     False
94     False
95     False
96      True
97      True
98     False
99      True
100     True
101    False
102    False

[103 rows x 1 columns]

In [40]:
df2= df1['borough'].str.contains('Toronto').to_frame()

In [41]:
df2.head()

borough
0    False
1    False
2     True
3    False
4     True

In [38]:
df2.drop('borough', axis = 1, inplace = True)

In [42]:
df2.head()

borough
0    False
1    False
2     True
3    False
4     True

In [44]:
df2 = df2.rename(columns = {'borough' :'tt'})

In [45]:
df1['temp'] = df2['tt']

In [46]:
df1.head()

postalcode           borough                                 neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude   temp  
0  43.753259 -79.329656  False  
1  43.725882 -79.315572  False  
2  43.654260 -79.360636   True  
3  43.718518 -79.464763  False  
4  43.662301 -79.389494   True

In [47]:
df1 = df1.drop(df1[(df1['temp'] == False)].index)

In [48]:
df1.head()

postalcode           borough                                 neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  temp  
2   43.654260 -79.360636  True  
4   43.662301 -79.389494  True  
9   43.657162 -79.378937  True  
15  43.651494 -79.375418  True  
19  43.676357 -79.293031  True

In [49]:
df1 = df1.reset_index()

In [50]:
df1.head()

index postalcode           borough  \
0      2        M5A  Downtown Toronto   
1      4        M7A  Downtown Toronto   
2      9        M5B  Downtown Toronto   
3     15        M5C  Downtown Toronto   
4     19        M4E      East Toronto   

                                  neighborhood   Latitude  Longitude  temp  
0                    Regent Park, Harbourfront  43.654260 -79.360636  True  
1  Queen's Park, Ontario Provincial Government  43.662301 -79.389494  True  
2                     Garden District, Ryerson  43.657162 -79.378937  True  
3                               St. James Town  43.651494 -79.375418  True  
4                                  The Beaches  43.676357 -79.293031  True

In [51]:
df1.drop('index', axis=1, inplace = True)

FINAL DATAFRAME CONTAINING TORONTO NEIGHBORHOODS ONLY

In [52]:
df1.head()

postalcode           borough                                 neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  temp  
0  43.654260 -79.360636  True  
1  43.662301 -79.389494  True  
2  43.657162 -79.378937  True  
3  43.651494 -79.375418  True  
4  43.676357 -79.293031  True

In [53]:
df1.shape

(39, 6)

GENERATE MAP to visualize TORONTO neighborhoods and how they cluster together

In [54]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

In [55]:
address = 'Toronto'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [56]:
import folium
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df1['Latitude'], df1['Longitude'], df1['neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto